### PR TITLE
Remove use of BouncyCastle for EC key generation and point (de)compression

### PR DESCRIPTION
**Pull request by WesleyRosenblum**
*Issue #, if available:* #41

*Description of changes:*

Removes explicit use of BouncyCastle from the `ECDSASignatureAlgorithm` implementation of `TrailingSignatureAlgorithm`.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

# Check any applicable:
- [ ] Were any files moved? Moving files changes their URL, which breaks all hyperlinks to the files.

### CODE REVIEW

File: src/main/java/com/amazonaws/encryptionsdk/internal/TrailingSignatureAlgorithm.java [hunk 1]
```diff
@@ -66,37 +83,99 @@ public String getRawSignatureAlgorithm() {
...
+                yOrder = ONE;
+            } else {
+                throw new IllegalArgumentException("Compressed y value was invalid");
+            }
+
+            final BigInteger p = ((ECFieldFp)ecParameterSpec.getCurve().getField()).getP();
+            final BigInteger a = ecParameterSpec.getCurve().getA();
+            final BigInteger b = ecParameterSpec.getCurve().getB();
+
+            final BigInteger alpha = x.modPow(THREE, p)
```
Comment (SalusaSecondus) [inline src/main/java/com/amazonaws/encryptionsdk/internal/TrailingSignatureAlgorithm.java]: Please add a comment making it clear that alpha is intended to be y^2 (and that we validate this later)

File: src/main/java/com/amazonaws/encryptionsdk/internal/Utils.java [hunk 2]
```diff
@@ -279,4 +280,30 @@ public static ByteBuffer limit(final ByteBuffer buff, final int newLimit) {
...
+
+        if(length < rawBytes.length) {
+            throw new IllegalArgumentException("Length must be as least as long as the BigInteger byte array " +
+                    "without the sign byte");
+        }
+
+        System.arraycopy(rawBytes, 0, result, length - rawBytes.length, rawBytes.length);
+
+        return result;
+    }
```
Comment (SalusaSecondus) [inline src/main/java/com/amazonaws/encryptionsdk/internal/Utils.java]: Let's simplify this (and reduce unneeded allocations).

```
final byte[] rawBytes = bigInteger.toByteArray();
// If we're already the correct length, just return it.
if (rawBytes.length ==length) {
    return rawBytes;
}
// If we're exactly one byte too large, but we have a leading zero byte, remove it and return.
if (rawBytes.length == length + 1 && rawBytes[0] == 0) {
    return Arrays.copyOfRange(rawBytes, 1, rawBytes.length);
}

if (rawBytes.length > length) {
    throw new IllegalArgumentException("Length must be as least as long as the BigInteger byte array " +
      "without the sign byte");
}
final byte[] paddedResult = new byte[length];
System.arraycopy(rawBytes, 0, paddedResult , length - rawBytes.length, rawBytes.length);
return paddedResult;
}
```

File: src/test/java/com/amazonaws/encryptionsdk/TestUtils.java [hunk 3]
```diff
@@ -174,4 +174,39 @@ public static ByteArrayInputStream insecureRandomStream(int length) {
...
+    /**
+     * Converts an array of unsigned bytes (represented as int values between 0 and 255 inclusive)
+     * to an array of Java primitive type byte, which are by definition signed.
+     * @param unsignedBytes An array on unsigned bytes
+     * @return An array of signed bytes
+     */
+  public static byte[] unsignedBytesToSignedBytes(final int[] unsignedBytes) {
+      byte[] signedBytes = new byte[unsignedBytes.length];
+
+      for(int i= 0 ; i < unsignedBytes.length; i++) {
```
Comment (SalusaSecondus) [inline src/test/java/com/amazonaws/encryptionsdk/TestUtils.java]: Here and throughout: space after `if`, `for`, and other keywords to distinguish them from function calls.

Here: Space on either side of the equals.

File: src/test/java/com/amazonaws/encryptionsdk/UtilsTest.java [hunk 4]
```diff
@@ -96,5 +97,19 @@ public void base64something() {
         assertEquals(encoded, Utils.encodeBase64String(data));
         assertArrayEquals(data, Utils.decodeBase64String(encoded));
     }
+
+    @Test
+    public void testBigIntegerToByteArray() {
```
Comment (SalusaSecondus) [inline src/test/java/com/amazonaws/encryptionsdk/UtilsTest.java]: Please also test exceptional cases.

File: src/main/java/com/amazonaws/encryptionsdk/internal/TrailingSignatureAlgorithm.java [hunk 5]
```diff
@@ -1,22 +1,18 @@
...
-import org.bouncycastle.jcajce.provider.asymmetric.ec.BCECPublicKey;
-import org.bouncycastle.jce.ECNamedCurveTable;
-import org.bouncycastle.jce.interfaces.ECPublicKey;
-import org.bouncycastle.jce.provider.BouncyCastleProvider;
-import org.bouncycastle.jce.spec.ECNamedCurveParameterSpec;
-import org.bouncycastle.math.ec.ECPoint;
+import java.math.BigInteger;
+import java.security.*;
+import java.security.interfaces.ECPublicKey;
+import java.security.spec.*;
```
[the remark this answers is not in the document]
Comment (WesleyRosenblum) [reply src/main/java/com/amazonaws/encryptionsdk/internal/TrailingSignatureAlgorithm.java]: Fixed, and updated my IDE

File: src/main/java/com/amazonaws/encryptionsdk/internal/TrailingSignatureAlgorithm.java [hunk 6]
```diff
@@ -1,22 +1,18 @@
...
+import java.util.Arrays;
 
 import com.amazonaws.encryptionsdk.CryptoAlgorithm;
 
-import static com.amazonaws.encryptionsdk.internal.BouncyCastleConfiguration.INTERNAL_BOUNCY_CASTLE_PROVIDER;
+import static com.amazonaws.encryptionsdk.internal.Utils.bigIntegerToByteArray;
+import static com.amazonaws.encryptionsdk.internal.Utils.encodeBase64String;
+import static org.apache.commons.lang3.Validate.*;
```
[the remark this answers is not in the document]
Comment (WesleyRosenblum) [reply src/main/java/com/amazonaws/encryptionsdk/internal/TrailingSignatureAlgorithm.java]: Fixed

File: src/main/java/com/amazonaws/encryptionsdk/internal/TrailingSignatureAlgorithm.java [hunk 7]
```diff
@@ -66,37 +83,99 @@ public String getRawSignatureAlgorithm() {
...
+
+            if(compressedY == TWO.byteValue()) {
+                yOrder = ZERO;
+            } else if (compressedY == THREE.byteValue()) {
+                yOrder = ONE;
+            } else {
+                throw new IllegalArgumentException("Compressed y value was invalid");
+            }
+
+            final BigInteger p = ((ECFieldFp)ecParameterSpec.getCurve().getField()).getP();
```
[the remark this answers is not in the document]
Comment (WesleyRosenblum) [reply src/main/java/com/amazonaws/encryptionsdk/internal/TrailingSignatureAlgorithm.java]: Done

File: src/main/java/com/amazonaws/encryptionsdk/internal/TrailingSignatureAlgorithm.java (hunk 1 above)
Comment (WesleyRosenblum) [reply src/main/java/com/amazonaws/encryptionsdk/internal/TrailingSignatureAlgorithm.java]: Added

File: src/main/java/com/amazonaws/encryptionsdk/internal/Utils.java (hunk 2 above)
Comment (WesleyRosenblum) [reply src/main/java/com/amazonaws/encryptionsdk/internal/Utils.java]: Done

File: src/test/java/com/amazonaws/encryptionsdk/TestUtils.java (hunk 3 above)
Comment (WesleyRosenblum) [reply src/test/java/com/amazonaws/encryptionsdk/TestUtils.java]: Fixed

File: src/test/java/com/amazonaws/encryptionsdk/UtilsTest.java (hunk 4 above)
Comment (WesleyRosenblum) [reply src/test/java/com/amazonaws/encryptionsdk/UtilsTest.java]: Added tests

**Pull request merged**